cpp/velox/shuffle/VeloxSortShuffleWriter.cc

```diff
@@ -0,0 +1,314 @@
+  initRowType(rv);
+  RETURN_NOT_OK(insert(rv, memLimit));
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortShuffleWriter::stop() {
+  ARROW_RETURN_IF(evictState_ == EvictState::kUnevictable, arrow::Status::Invalid("Unevictable state in stop."));
+
+  EvictGuard evictGuard{evictState_};
+
+  stopped_ = true;
+  RETURN_NOT_OK(evictAllPartitions());
+  RETURN_NOT_OK(partitionWriter_->stop(&metrics_));
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortShuffleWriter::reclaimFixedSize(int64_t size, int64_t* actual) {
+  if (evictState_ == EvictState::kUnevictable || offset_ == 0) {
+    *actual = 0;
+    return arrow::Status::OK();
+  }
+  EvictGuard evictGuard{evictState_};
+  auto beforeReclaim = veloxPool_->usedBytes();
+  RETURN_NOT_OK(evictAllPartitions());
+  *actual = beforeReclaim - veloxPool_->usedBytes();
+  return arrow::Status::OK();
+}
+
+arrow::Status VeloxSortShuffleWriter::init() {
+  ARROW_RETURN_IF(
+      options_.partitioning == Partitioning::kSingle,
+      arrow::Status::Invalid("VeloxSortShuffleWriter doesn't support single partition."));
+  array_.resize(initialSize_);
+  return arrow::Status::OK();
+}
+
+void VeloxSortShuffleWriter::initRowType(const facebook::velox::RowVectorPtr& rv) {
+  if (UNLIKELY(!rowType_)) {
+    rowType_ = facebook::velox::asRowType(rv->type());
+    fixedRowSize_ = facebook::velox::row::CompactRow::fixedRowSize(rowType_);
+  }
+}
+
+arrow::Result<facebook::velox::RowVectorPtr> VeloxSortShuffleWriter::getPeeledRowVector(
+    const std::shared_ptr<ColumnarBatch>& cb) {
+  if (options_.partitioning == Partitioning::kRange) {
+    auto compositeBatch = std::dynamic_pointer_cast<CompositeColumnarBatch>(cb);
+    VELOX_CHECK_NOT_NULL(compositeBatch);
+    auto batches = compositeBatch->getBatches();
+    VELOX_CHECK_EQ(batches.size(), 2);
+
+    auto pidBatch = VeloxColumnarBatch::from(veloxPool_.get(), batches[0]);
+    auto pidArr = getFirstColumn(*(pidBatch->getRowVector()));
+    RETURN_NOT_OK(partitioner_->compute(pidArr, pidBatch->numRows(), row2Partition_));
+
+    auto rvBatch = VeloxColumnarBatch::from(veloxPool_.get(), batches[1]);
+    return rvBatch->getFlattenedRowVector();
+  } else {
+    auto veloxColumnBatch = VeloxColumnarBatch::from(veloxPool_.get(), cb);
+    VELOX_CHECK_NOT_NULL(veloxColumnBatch);
+    auto rv = veloxColumnBatch->getFlattenedRowVector();
+    if (partitioner_->hasPid()) {
+      auto pidArr = getFirstColumn(*rv);
+      RETURN_NOT_OK(partitioner_->compute(pidArr, rv->size(), row2Partition_));
+      return getStrippedRowVector(*rv);
+    } else {
+      RETURN_NOT_OK(partitioner_->compute(nullptr, rv->size(), row2Partition_));
+      return rv;
+    }
+  }
+}
+
+arrow::Status VeloxSortShuffleWriter::insert(const facebook::velox::RowVectorPtr& vector, int64_t memLimit) {
+  ScopedTimer timer(&c2rTime_);
+  auto inputRows = vector->size();
+  VELOX_DCHECK_GT(inputRows, 0);
+
+  facebook::velox::row::CompactRow row(vector);
```

Comment:
Hey, @marin-ma , might be a stupid question, why we don't use compact row for rss shuffle, I see VeloxCelebornSortShuffleWriter still use PrestoPage to do serialization, is there any reason why we don't change them as well. Does PrestoPage have better performance than Compact row when doing serialization?

Reply:
PrestoSerializer still serialises the data by column. We use compact row for sort based shuffle is that we found the performance drops significantly with large shuffle partitions and wide tables for both using hash-based shuffle or presto-serialiser.